services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/validation/HankeValidatorTest.kt

```diff
@@ -0,0 +1,275 @@
+
+        verifyError(HankeError.HAI1002, "vaihe")
+    }
+
+    @Test
+    fun `fails if vaihe is suunnittelu and suunnitteluVaihe is null`() {
+        val hanke = HankeFactory.create(vaihe = Vaihe.SUUNNITTELU, suunnitteluVaihe = null)
+
+        assertThat(hankeValidator.isValid(hanke, context)).isFalse()
+
+        verifyError(HankeError.HAI1002, "suunnitteluVaihe")
+    }
+
+    @Test
+    fun `succeeds if vaihe is suunnittelu and suunnitteluVaihe is defined`() {
+        val hanke =
+            HankeFactory.create(
+                vaihe = Vaihe.SUUNNITTELU,
+                suunnitteluVaihe = SuunnitteluVaihe.YLEIS_TAI_HANKE
+            )
+
+        assertThat(hankeValidator.isValid(hanke, context)).isTrue()
+    }
+
+    @Test
+    fun `succeeds if tyomaaKatuosoite is null`() {
+        val hanke = HankeFactory.create().apply { tyomaaKatuosoite = null }
+
+        assertThat(hankeValidator.isValid(hanke, context)).isTrue()
+    }
+
+    @Test
+    fun `succeeds if tyomaaKatuosoite is almost too long`() {
+        val hanke =
+            HankeFactory.create().apply {
+                tyomaaKatuosoite = "F".repeat(MAXIMUM_TYOMAAKATUOSOITE_LENGTH)
+            }
+
+        assertThat(hankeValidator.isValid(hanke, context)).isTrue()
+    }
+
+    @Test
+    fun `fails if tyomaaKatuosoite is too long`() {
+        val hanke =
+            HankeFactory.create().apply {
+                tyomaaKatuosoite = "F".repeat(MAXIMUM_TYOMAAKATUOSOITE_LENGTH + 1)
+            }
+
+        assertThat(hankeValidator.isValid(hanke, context)).isFalse()
+
+        verifyError(HankeError.HAI1002, "tyomaaKatuosoite")
+    }
+
+    @Test
+    fun `succeeds for the default hanke with an alue`() {
+        val hanke = HankeFactory.create().withHankealue()
+
+        assertThat(hankeValidator.isValid(hanke, context)).isTrue()
+    }
+
+    @Test
+    fun `succeeds if hanke alue nimi is null`() {
+        val hanke = HankeFactory.create().withHankealue(nimi = null)
+
+        assertThat(hankeValidator.isValid(hanke, context)).isTrue()
+    }
+
+    @Test
+    fun `succeeds if hanke alue nimi is almost too long`() {
+        val hanke =
+            HankeFactory.create().withHankealue(nimi = "F".repeat(MAXIMUM_HANKE_ALUE_NIMI_LENGTH))
+
+        assertThat(hankeValidator.isValid(hanke, context)).isTrue()
+    }
+
+    @Test
+    fun `fails if hanke alue nimi is null`() {
+        val hanke =
+            HankeFactory.create()
+                .withHankealue(nimi = "F".repeat(MAXIMUM_HANKE_ALUE_NIMI_LENGTH + 1))
+
+        assertThat(hankeValidator.isValid(hanke, context)).isFalse()
+
+        verifyError(HankeError.HAI1032, "alueet[0].nimi")
+    }
+
+    @Test
+    fun `fails if haitta alku pvm is too far in the future`() {
+        val hanke = HankeFactory.create().withHankealue(haittaAlkuPvm = MAXIMUM_DATE.plusDays(1))
+
+        assertThat(hankeValidator.isValid(hanke, context)).isFalse()
+
+        verifyError(HankeError.HAI1032, "alueet[0].haittaAlkuPvm")
```

Comment:
Sonar is complaining about "Define a constant instead of duplicating this literal".

I'm not sure if it is possible for merge from dev to main will fail code quality if we have too many of these from different individual commits. So I guess they should be fixed?

Reply:
I resolved the issues in Sonarcloud. I think using a constant here would be very distracting.